packages/xod-client-electron/src/app/arduinoActions.js

```diff
@@ -213,7 +204,100 @@ export const uploadToArduinoHandler = (event, payload) => {
     }
-  )(payload).catch(convertAndSendError);
+  )(payload);
+};
+
+// =============================================================================
+//
+// Cloud compiling and uploading case
+//
+// =============================================================================
+
+// :: UploadConfig -> Promise Path Error
+const installTool = R.converge(
+  xd.installTool(uploadToolsPath),
+  [
+    xd.getToolVersionPath,
+    xd.getToolUrl,
+  ]
+);
+
+const deployToArduinoThroughCloud = ({
+  payload,
+  sendProgress,
+  sendSuccess,
+}) => R.pipeP(
+  doTranspileForArduino,
+  sendProgress(MESSAGES.CODE_TRANSPILED, 10),
+  code => checkPort(payload.port).then(port => ({ code, port: port.comName })),
+  sendProgress(MESSAGES.PORT_FOUND, 15),
+  ({ code, port }) => ({ code, port, pio: payload.board.pio }),
+  // ({ code, port, pio }) => uploadToArduino(xad.strigifyFQBN(pab), port, code),
+  data => xd.getUploadConfig(data.pio)
+    .then(cfg => R.assoc('uploadConfig', cfg, data))
+    .catch((err) => {
+      const errCode = R.cond([
+        [R.propEq('status', 404), R.always(ERROR_CODES.BOARD_NOT_SUPPORTED)],
+        [R.has('status'), R.always(ERROR_CODES.CANT_GET_UPLOAD_CONFIG)],
+        [R.T, R.always(ERROR_CODES.CLOUD_NETWORK_ERROR)],
+      ])(err);
+
+      return rejectWithCode(errCode, err);
+    }),
+  data => installTool(data.uploadConfig)
+    .then(res => R.assoc('toolPath', res.path, data)),
+  sendProgress(MESSAGES.CLOUD_TOOLCHAIN_INSTALLED, 30),
+  data => xd.compile(data.pio, data.code)
+    .then(xd.saveCompiledBinary(artifactTmpDir))
+    .then(artifactPath => R.assoc('artifactPath', artifactPath, data)),
+  sendProgress(MESSAGES.CODE_COMPILED, 75),
+  ({ uploadConfig, toolPath, artifactPath, port }) =>
+    xd.upload(uploadConfig, { toolPath, artifactPath, port }),
+  (result) => {
+    if (result.exitCode !== 0) {
+      return Promise.reject(Object.assign(new Error(`Upload tool exited with error code: ${result.exitCode}`), result));
+    }
+    return sendSuccess([result.stderr, result.stdout].join('\n\n'), 100)();
+  }
+)(payload);
+
+// =============================================================================
+//
+// IPC handlers (for main process)
+//
+// =============================================================================
+
+export const uploadToArduinoHandler = (event, payload) => {
+  // Messages
+  const send = status => R.compose(
+    data => (arg) => { event.sender.send('UPLOAD_TO_ARDUINO', data); return arg; },
+    R.assoc(status, true),
+    (message, percentage, err = null) => ({
+      success: false,
+      progress: false,
+      failure: false,
+      error: err,
+      message,
+      percentage,
+    })
+  );
+  const sendSuccess = send('success');
+  const sendProgress = send('progress');
+  const sendFailure = send('failure');
+  const convertAndSendError = err => R.compose(
+    msg => sendFailure(msg, 0, errorToPlainObject(err))(),
```

Comment:
Do we really need to reset progress bar to 0 on error?
Maybe, it's useful to show user at which stage where an error has happened?